fix(terminal): don't require ! to close a stopped terminal buffer

gpanders opened the pull request:
Closes #4683.

- If the job in a terminal buffer is still running then `!` is still required.
- Change jobwait() such that it always flushes the proc-events queue. Previously it only did this when it actually waited on the process.

gpanders committed:
fix(terminal): close without ! if the job is stopped

- If the terminal job is still running then ! is still required.

Closes #4683

justinmk commented:
Looking forward to this, but definitely needs a test. `test/functional/terminal/buffer_spec.lua`

justinmk commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -255,8 +256,11 @@ describe(':terminal buffer', function()
     eq('Vim(wqall):E948: Job still running', exc_exec('wqall'))
+    command('call jobstop(&channel)')
+    retry(nil, nil, function() eq(-3, eval('jobwait([&channel], 0)[0]')) end)
+    command('bdelete')
```
windows ci

```
[  ERROR   ] test/functional\terminal\buffer_spec.lua @ 259: :terminal buffer won't close while a job is running (without !)
  test\functional\helpers.lua:113: Vim(bdelete):E89: term://D:\a\neovim\neovim//1556:D:/a/neovim/neovim/build/bin/tty-test.EXE will be killed (add ! to override)

gpanders replied:
I'm not sure why this is failing. If it makes it past the `retry` (meaning `jobwait()` returns `-3`) the job should be stopped...

gpanders replied:
There's a race condition somewhere, as even running the test locally fails randomly.

gpanders replied:
Using `feed_command` instead of `command` seems to fix it. My best guess is that `command` was being processed before the event loop called `process_close_event`.

justinmk replied:
not sure we should accept that behavior. jobwait() should be reliable, otherwise plugin authors have to work around our bugs...

gpanders replied:
I think I was wrong about the race condition. `f_jobwait` processes the job's event queue after the `process_wait` calls `decref` which enqueues `process_close_event`:

https://github.com/gpanders/neovim/blob/1ab8ead3a3efff0fca82e7cbdaef5648c235d28d/src/nvim/eval/funcs.c#L5414

But this leaves me back at square one. I'm no longer able to reproduce the failure on my dev machine, and it seems like it only fails on Windows, but I have no ideas why that would be.

justinmk replied:
> And _even if_ the job still is valid, `f_jobwait` passes a different `MultiQueue` object to `process_wait`, so the `events` argument in the snippet above doesn't actually belong to `proc`

Nice catch, this looks like a mistake from f1de097dbb236ea400150f80b909407ca9af7441 .

gpanders replied:
I put that particular change in a separate PR #15501 as I don't think that's our problem in this case. As I mentioned, `f_jobwait` isn't even always hitting `process_wait` since in a lot of cases the job is already stopped by the time we enter `f_jobwait`.

Checking the job directly [is implemented in this PR](https://github.com/neovim/neovim/pull/15402/files#diff-0b5b9677295695ef4fc7ce3c1a86a30c0c5633c5275ff1424a4e4219e01fa3fbR1241-R1245), though now the Windows CI is failing for a different reason :(

justinmk replied:
> `f_jobwait` passes a different `MultiQueue` object to `process_wait`, so the `events` argument in the snippet above doesn't actually belong to `proc`.

Just to be clear (you probably know already): that "only" affects ordering, since `proc->events` is a child of `main_loop.events`.

justinmk commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -255,8 +256,11 @@ describe(':terminal buffer', function()
   end)
 
-  it('handles wqall', function()
+  it("won't close while a job is running (without !)", function()
```
```suggestion
  it("requires bang (!) to close a running job", function()
```

justinmk commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -255,8 +256,11 @@ describe(':terminal buffer', function()
 
-  it('handles wqall', function()
+  it("won't close while a job is running (without !)", function()
     eq('Vim(wqall):E948: Job still running', exc_exec('wqall'))
```
would also add asserts for :bdelete and :quit here. Their implementations are all slightly different.

gpanders replied:
`:quit` already works without any warning or error, which is probably a bug.

justinmk replied:
fixing the inverse is a different topic i guess, but we should still add the :quit case here.

gpanders replied:
How should it be added though? `:quit` doesn't throw an exception, it simply terminates the running Nvim process as well as the terminal process. What are we testing there?

justinmk replied:
It at least is a regression test (though it should go at the end of the test, of course, since the test nvim will exit), which adds coverage/when we fix :quit to warn when the terminal-job is running.

justinmk commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -255,8 +261,14 @@ describe(':terminal buffer', function()
+    matches('^Vim%(bdelete%):E89: term://'..pesc(cwd)..'//%d+:'..nvim_dir..'/tty%-test'..ext_pat..' will be killed %(add %! to override%)$', exc_exec('bdelete'))
+    command('call jobstop(&channel)')
+    retry(nil, nil, function() neq(-1, eval('jobwait([&channel], 0)[0]')) end)
+    feed_command('bdelete')
```
feed_command is async, the test won't wait and any errors will be ignored. More details at `:help nvim_input`

justinmk commented on src/nvim/buffer.c:
```diff
@@ -1231,20 +1231,25 @@ do_buffer(
+
+    if (!forceit && buf->terminal) {
+      uint64_t chan_id;
+      get_option_value("channel", (long *)&chan_id, NULL, OPT_LOCAL);
```
this makes an assumption about curbuf which we shouldn't introduce here. I was hoping that `f_getbufvar` would lead to a function that takes a `buf`, but no... https://github.com/neovim/neovim/blob/8af13ed946aa06636c633327c6549993e2e50116/src/nvim/eval/funcs.c#L3004-L3024

gpanders replied:
We can use `get_option_value_strict` instead which takes a `buf_T` as its `from` parameter (this is how `nvim_buf_get_option` does it).

justinmk commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -255,8 +261,14 @@ describe(':terminal buffer', function()
     eq('Vim(wqall):E948: Job still running', exc_exec('wqall'))
+    matches('^Vim%(bdelete%):E89: term://'..pesc(cwd)..'//%d+:'..nvim_dir..'/tty%-test'..ext_pat..' will be killed %(add %! to override%)$', exc_exec('bdelete'))
+    command('call jobstop(&channel)')
+    retry(nil, nil, function() neq(-1, eval('jobwait([&channel], 0)[0]')) end)
```
-1 doesn't seem like the right thing to check. `:help jobwait`

> -1 if the timeout was exceeded
> -2 if the job was interrupted (by |CTRL-C|)
> -3 if the job-id is invalid

thus we should check eq(-3, ...)
```suggestion
    retry(nil, nil, function() eq(-3, eval('jobwait([&channel], 0)[0]')) end)
```

justinmk replied:
a less spammy way would be to pass a timeout to jobwait(), instead of calling it zillions of times with timeout=0.
```suggestion
    retry(nil, nil, function() assert(0 <= eval('jobwait([&channel], 1000)[0]')) end)
```

justinmk commented:
nice explanation in 3f42c600fec40e3619b239caf20485199a6dcfce. however, stepping back a bit...

> A terminal buffer is only marked closed after the process's exit callback is processed on the event loop. It's possible that this won't happen until after an attempt to close the terminal buffer, which then fails since the terminal is not yet marked as closed. However, closing the buffer should succeed so long as the underlying process has indeed exited.

Is the "problem" that `f_jobstop` / `process_stop` are asynchronous? The usual and expected pattern is to check `jobwait()` (or some other way to check the job's status). 

If `jobwait()` is lying to us then we need to fix that. Doesn't seem like `:bdelete` should have to care about various details of the event loop / job subystem.

> see the test in terminal/buffer_spec which calls 'bdelete' immediately after 'jobstop'. In this case, 'bdelete' is processed before the terminal is marked closed, even though the process has terminated.

`jobstop` doesn't guarantee that. `jobwait` however should and must guarantee that the job really is finished. If jobwait() says the job is dead then `:bdelete` and anything else must be able to conclude that the job is dead.

gpanders commented:
> Is the "problem" that `f_jobstop` / `process_stop` are asynchronous? The usual and expected pattern is to check `jobwait()` (or some other way to check the job's status).
> 
> If `jobwait()` is lying to us then we need to fix that. Doesn't seem like `:bdelete` should have to care about various details of the event loop / job subystem.
> 
> > see the test in terminal/buffer_spec which calls 'bdelete' immediately after 'jobstop'. In this case, 'bdelete' is processed before the terminal is marked closed, even though the process has terminated.
> 
> `jobstop` doesn't guarantee that. `jobwait` however should and must guarantee that the job really is finished. If jobwait() says the job is dead then `:bdelete` and anything else must be able to conclude that the job is dead.

I don't think `jobwait()` is lying, but there are multiple paths it can take:

1) The provided job is already stopped (return `-3`)
2) The provided job is still running, and the timeout is set to 0 (return `-1`)
3) The provided job is still running, and the timeout is non-zero (call `process_wait`)

(3) is the only one that calls `process_wait`, so that is the only code path that will in turn immediately call `channel_process_exit_cb` (which is where `term->closed` is set to `true`). Otherwise that callback isn't executed until the process is fully dereferenced, which may not happen until the event loop spins again.

So `jobwait` *does* guarantee that the process is terminated, but not that all of the event loop callbacks have fired. Maybe the solution is that `jobwait` should also drain all of the event queues of the processes that are passed to it?

justinmk commented:
> (3) is the only one that calls `process_wait`, so that is the only code path that will in turn immediately call `channel_process_exit_cb` (which is where `term->closed` is set to `true`). Otherwise that callback isn't executed until the process is fully dereferenced, which may not happen until the event loop spins again.
> 
> So `jobwait` _does_ guarantee that the process is terminated, but not that all of the event loop callbacks have fired. Maybe the solution is that `jobwait` should also drain all of the event queues of the processes that are passed to it?

I would say so. If it saves code maybe it can just call process_wait in all cases.

gpanders committed:
fix(jobwait): always drain process event queues #15402

Problem:
jobwait() returns early if the job was stopped, but the job might have
pending callbacks on its event queue which are required to complete its
teardown. State such as term->closed might not be updated yet (by the
pending callbacks), so codepaths such as :bdelete think the job is still
running.

Solution:
Always flush the job's event queue before returning from jobwait().

ref #15349

gpanders commented:
All tests passing! (lint check fails on `screen.c` which is unrelated)

jamessan commented:
> So `jobwait` _does_ guarantee that the process is terminated, but not that all of the event loop callbacks have fired. Maybe the solution is that `jobwait` should also drain all of the event queues of the processes that are passed to it?

Hmm, does this also potentially fix https://github.com/neovim/neovim/issues/15349?

justinmk commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -255,8 +261,21 @@ describe(':terminal buffer', function()
     eq('Vim(wqall):E948: Job still running', exc_exec('wqall'))
+    matches('^Vim%(bdelete%):E89: term://'..pesc(cwd)..'//%d+:'..nvim_dir..'/tty%-test'..ext_pat..' will be killed %(add %! to override%)$', exc_exec('bdelete'))
+    command('call jobstop(&channel)')
+    retry(nil, nil, function() assert(0 >= eval('jobwait([&channel], 1000)[0]')) end)
```
because jobwait() is actually waiting, we don't need retry() now.
```suggestion
    assert(0 >= eval('jobwait([&channel], 1000)[0]'))
```

justinmk commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -255,8 +261,21 @@ describe(':terminal buffer', function()
+  it('stops running jobs with :quit', function()
+    -- Open in a new window to avoid terminating the nvim instance
+    command('split')
+    command('terminal')
```
this test needs "set nohidden" so let's be explict
```suggestion
    command('terminal')
    command('set nohidden')
```

justinmk commented on src/nvim/eval/funcs.c:
```diff
@@ -5404,8 +5404,17 @@ static void f_jobwait(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+      Process *proc = &chan->stream.proc;
+      LOOP_PROCESS_EVENTS_UNTIL(proc->loop, proc->events, -1, !proc->refcount);
+
+      // Set job to NULL so that the return value will be set to -3 below
+      jobs[i] = NULL;
```
as a reader it would help me see a pattern if this stays identical with the case above. the longer comment doesn't seem to add anything.
```suggestion
      jobs[i] = NULL;  // Invalid job.
```

gpanders replied:
I avoided that language because I didn’t think a stopped job was “invalid”, but I also don’t feel strongly about it and I certainly want to improve readability if possible. So this is fine with me.

justinmk commented on src/nvim/eval/funcs.c:
```diff
@@ -5404,8 +5404,17 @@ static void f_jobwait(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+      // Process is already stopped, so ensure all callbacks on its event
+      // queue are drained
+      Process *proc = &chan->stream.proc;
+      LOOP_PROCESS_EVENTS_UNTIL(proc->loop, proc->events, -1, !proc->refcount);
```
- any reason not to use `process_wait()`? that avoids reaching into `proc->refcount` (informally private), and also fits semantically, so a reader can see that process_wait() is called in multiple places in jobwait().

```suggestion
      process_wait(&chan->stream.proc, -1, NULL);
```

also via @bfredl:

> it should be documented that any object holding a proc->refcount must release that eventually by an event on the process own event queue. and i.e. not by some event on some other child queue (only)

gpanders replied:
The idea was to cut to the chase and sidestep the other stuff that process_wait does, as well as being explicit that we are just processing the event queue and not necessarily “waiting” for a process that has already stopped.

But I think you’re right that it may be clearer. And I wasn’t aware of the semantics around `refcount`, so I think your suggestion makes sense.

(Btw I’m replying on mobile and just using the “Commit suggestion” box, so forgive the lack of better commit messages. I figure they can be squashed/rebased later).

justinmk commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -255,8 +260,22 @@ describe(':terminal buffer', function()
+    local cwd = funcs.fnamemodify('.', ':p:~'):gsub([[[\/]*$]], '')
+    local ext_pat = iswin() and '%.EXE' or ''
     eq('Vim(wqall):E948: Job still running', exc_exec('wqall'))
+    matches('^Vim%(bdelete%):E89: term://'..pesc(cwd)..'//%d+:'..nvim_dir..'/tty%-test'..ext_pat..' will be killed %(add %! to override%)$', exc_exec('bdelete'))
```
- we could simplify the test by not being so precise. I don't think we lose much by changing the pattern to ignore the CWD and EXE. Up to you, though.
    - this anticipates a glorious future where we change the `term://` format https://github.com/neovim/neovim/issues/3278

```suggestion
    matches('^Vim%(bdelete%):E89: term://.*tty%-test.* will be killed %(add %! to override%)$', exc_exec('bdelete'))
```

gpanders replied:
No objection to this.

justinmk commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -255,8 +260,22 @@ describe(':terminal buffer', function()
-  it('handles wqall', function()
+  it("requires bang (!) to close a running job", function()
+    local cwd = funcs.fnamemodify('.', ':p:~'):gsub([[[\/]*$]], '')
+    local ext_pat = iswin() and '%.EXE' or ''
```
ext_pat is no longer used (helpers.iswin and other imports can be removed also)

justinmk commented:
Squashed some commits and pushed some extra test coverage for `:bunload`, `:bwipeout`.
